src/main/kotlin/com/exactpro/th2/read/file/common/impl/DefaultFileReader.kt

```diff
@@ -77,122 +84,305 @@ class DefaultFileReader<T : AutoCloseable> private constructor(
         delegateHolder.onSourceClosed.invoke(streamId, path)
     }
 
-    private data class DelegateHolder<T : AutoCloseable>(
+    private data class DelegateHolder<T : AutoCloseable, MESSAGE_BUILDER>(
```

Comment:
If we apply the suggested change regarding `createInstance` method in Builder this class should be protected

Reply:
No, it should be 'public' in this case because it will be exposed as parameter of 'createInstance' method (it is 'protected' in Builder, but 'Builder' itself is public in 'DefaultFileReader').

Reply:
@lumber1000 Sorry, but I don't get your point...
The constructor of `DefaultFileReader` has a protected modifier. The classes `ProtoDefaultFileReader` and `TransportDefaultFileReader` have private constructors. The builder uses `DelegateHolder` as a private field
```kotlin
private var delegateHolder = DelegateHolder<T, MESSAGE_BUILDER>()
```
And I don't really see the problem that you are describing...
Could you please elaborate on your position here?

Reply:
these changes causes compile time errors: `DefaultFileReader.kt:179:47 protected (in Builder)' function exposes its /protected (in DefaultFileReader)' parameter type DelegateHolder`. (and the same ProtoFileReader and TransportFileReader).

Why this happens: builder uses `DelegateHolder` not only as private field but also as a parameter in non-private method ('createInstance'). We can subclass `DefaultFileReader.Builder` from the outside of `DefaultFileReader` and this subclass will not be nested into `DefaultFileReader` and will have `createInstance` method with `DelegateHolder` (which should not be used outside of `DefaultFileReader` or it's subclasses).

Reply:
Thank you for the explanation. Now I get the problem. In this case, we can keep the `DelegateHolder` and the `lateinit` property in the `DefaultFileReader` class...